src/clib/pio_msg.c

```diff
@@ -1600,80 +1680,88 @@ int pio_msg_handler(int io_rank, int component_count, iosystem_desc_t *iosys)
 	    freedecomp_handler(my_iosys);
 	    break;
 	case PIO_MSG_EXIT:
-	    finalize_handler(my_iosys);
+	    finalize_handler(my_iosys, index);
 	    msg = -1;
 	    break;
 	default:
 	    pio_callback_handler(my_iosys, msg);
 	}
 
 	/* If an error was returned by the handler, do something! */
-	LOG((3, "pio_msg_handler checking error ret = %d", ret));
+	LOG((3, "pio_msg_handler2 checking error ret = %d", ret));
 	if (ret)
-	{
-	    LOG((0, "hander returned error code %d", ret));
 	    MPI_Finalize();
-	}
 
-	LOG((3, "pio_msg_handler getting ready to listen"));
+	LOG((3, "pio_msg_handler2 getting ready to listen"));
+
 	/* Unless finalize was called, listen for another msg from the
 	 * component whose message we just handled. */
 	if (!io_rank && msg != -1)
 	{
-	    LOG((3, "pio_msg_handler about to Irecv"));
-	    my_iosys = &iosys[index];
+	    my_iosys = iosys[index];
+	    LOG((3, "pio_msg_handler2 about to Irecv index = %d comproot = %d union_comm = %d",
+		 index, my_iosys->comproot, my_iosys->union_comm));
 	    mpierr = MPI_Irecv(&msg, 1, MPI_INT, my_iosys->comproot, MPI_ANY_TAG, my_iosys->union_comm,
 			       &req[index]);
-	    LOG((3, "pio_msg_handler called MPI_Irecv req[%d] = %d\n", index, req[index]));
+	    LOG((3, "pio_msg_handler2 called MPI_Irecv req[%d] = %d\n", index, req[index]));
 	    CheckMPIReturn(mpierr, __FILE__, __LINE__);
 	}
 
+	LOG((3, "pio_msg_handler2 done msg = %d open_components = %d",
+	     msg, open_components));
+
+	/* If there are no more open components, exit. */
+	if (msg == -1)
+	    if (--open_components)
+		msg = PIO_MSG_EXIT;
     }
 
+    LOG((3, "returning from pio_msg_handler2"));
     return PIO_NOERR;
 }
 
-int
-pio_iosys_print(int my_rank, iosystem_desc_t *iosys)
-{
-    printf("%d iosysid: %d\n", my_rank, iosys->iosysid);
-    if (iosys->union_comm == MPI_COMM_NULL)
-	printf("%d union_comm: MPI_COMM_NULL ", my_rank);
-    else
-	printf("%d union_comm: %d ", my_rank, iosys->union_comm);
-
-    if (iosys->comp_comm == MPI_COMM_NULL)
-	printf("comp_comm: MPI_COMM_NULL ");
-    else
-	printf("comp_comm: %d ", iosys->comp_comm);
-
-    if (iosys->io_comm == MPI_COMM_NULL)
-	printf("io_comm: MPI_COMM_NULL ");
-    else
-	printf("io_comm: %d ", iosys->io_comm);
-
-    if (iosys->intercomm == MPI_COMM_NULL)
-	printf("intercomm: MPI_COMM_NULL\n");
-    else
-	printf("intercomm: %d\n", iosys->intercomm);
-
-    printf("%d num_iotasks=%d num_comptasks=%d union_rank=%d, comp_rank=%d, "
-	   "io_rank=%d async_interface=%d\n",
-	   my_rank, iosys->num_iotasks, iosys->num_comptasks, iosys->union_rank,
-	   iosys->comp_rank, iosys->io_rank, iosys->async_interface);
-
-    printf("%d ioroot=%d comproot=%d iomaster=%d, compmaster=%d\n",
-	   my_rank, iosys->ioroot, iosys->comproot, iosys->iomaster,
-	   iosys->compmaster);
-
-    printf("%d iotasks:", my_rank);
-    for (int i = 0; i < iosys->num_iotasks; i++)
-	printf("%d ", iosys->ioranks[i]);
-    printf("\n");
-    return PIO_NOERR;
-}
+/* int */
```

Comment:
Yes, I will remove this commented-out function. I wrote it before I put LOG into the library. Now that we have LOGs I don't use this function anymore anyway.